src/response/file.rs

```diff
@@ -0,0 +1,92 @@
+use http::header::{CONTENT_LENGTH, CONTENT_TYPE, ETAG, LAST_MODIFIED};
+use httpdate::HttpDate;
+use std::fs::Metadata;
+use std::io::{self, ErrorKind};
+use std::path::{Path, PathBuf};
+use tokio::fs;
+
+use super::Response;
+use crate::{middleware, Error};
+
+type GenerateEtag = fn(&Metadata) -> Result<String, Error>;
+
+/// Serve a single file from disk.
+///
+pub struct File {
```

Comment:
I'm considering using references for all of these variables but want to see where the await points are before assuming things can be borrowed.

I would really like to use `&'a dyn FnOnce(&Metadata) -> Result<String, Error>` for the generate etag struct field but think that would require an additional `Box` and these seem like a lot of unnecessary allocations. It's important that this is sound from an async perspective but I'm going to try to come up with a better way. Using a function pointer may feel gross but it's likely our best path forward.

Reply:
In other words, if an attacker can overwrite the value of the function pointer between await points to a nefarious function of their own creation–we have bigger problems to worry about.